derive_client/_bridge/client.py

```diff
@@ -264,11 +274,19 @@ def _resolve_socket_route(
 
     async def _prepare_tx(
         self,
-        func: ContractFunction,
+        amount: int,
+        func: AsyncContractFunction,
         value: int,
         context: BridgeContext,
     ) -> PreparedBridgeTx:
 
+        onchain_decimals: int = await context.source_token.functions.decimals().call()
+        if onchain_decimals != (expected_decimals := CURRENCY_DECIMALS[context.currency]):
+            raise RuntimeError(
+                f"Decimal mismatch for {context.currency.name} on {context.source_chain.name}: "
+                f"expected {expected_decimals}, got {onchain_decimals}"
+            )
+
```

Comment:
Just as we verify the wallet owner before preparing a transaction, we also confirm the token decimals on-chain. This is necessary because we do not check every contract in the production addresses on every chain, and many tokens are proxies whose decimals could differ from our assumption